hack/generator/pkg/jsonast/schema_abstraction_openapi.go

```diff
@@ -0,0 +1,337 @@
+}
+
+func (it *OpenAPISchema) HasAnyOf() bool {
+	return len(it.schema.AnyOf) > 0
+}
+
+func (it *OpenAPISchema) AnyOf() []Schema {
+	return it.transformOpenAPISlice(it.schema.AnyOf)
+}
+
+func (it *OpenAPISchema) HasOneOf() bool {
+	return len(it.schema.OneOf) > 0
+}
+
+func (it *OpenAPISchema) OneOf() []Schema {
+	return it.transformOpenAPISlice(it.schema.OneOf)
+}
+
+func (it *OpenAPISchema) RequiredProperties() []string {
+	return it.schema.Required
+}
+
+func (it *OpenAPISchema) Properties() map[string]Schema {
+	result := make(map[string]Schema)
+	for propName, propSchema := range it.schema.Properties {
+		result[propName] = it.withNewSchema(propSchema)
+	}
+
+	return result
+}
+
+func (it *OpenAPISchema) Description() *string {
+	if len(it.schema.Description) == 0 {
+		return nil
+	}
+
+	return &it.schema.Description
+}
+
+func (it *OpenAPISchema) Items() []Schema {
+	if it.schema.Items.Schema != nil {
+		return []Schema{it.withNewSchema(*it.schema.Items.Schema)}
+	}
+
+	return it.transformOpenAPISlice(it.schema.Items.Schemas)
+}
+
+func (it *OpenAPISchema) AdditionalPropertiesAllowed() bool {
+	return it.schema.AdditionalProperties == nil || it.schema.AdditionalProperties.Allows
+}
+
+func (it *OpenAPISchema) AdditionalPropertiesSchema() Schema {
+	if it.schema.AdditionalProperties == nil {
+		return nil
+	}
+
+	result := it.schema.AdditionalProperties.Schema
+	if result == nil {
+		return nil
+	}
+
+	return it.withNewSchema(*result)
+}
+
+func (it *OpenAPISchema) EnumValues() []string {
+	result := make([]string, len(it.schema.Enum))
+	for i, enumValue := range it.schema.Enum {
+		if enumString, ok := enumValue.(string); ok {
+			result[i] = fmt.Sprintf("%q", enumString)
+		} else if enumStringer, ok := enumValue.(fmt.Stringer); ok {
+			result[i] = fmt.Sprintf("%q", enumStringer.String())
+		} else if enumFloat, ok := enumValue.(float64); ok {
+			result[i] = fmt.Sprintf("%g", enumFloat)
+		} else {
+			panic(fmt.Sprintf("unable to convert enum value (%v %T) to string", enumValue, enumValue))
+		}
+	}
+
+	return result
+}
+
+func (it *OpenAPISchema) IsRef() bool {
+	return it.schema.Ref.GetURL() != nil
+}
+
+type fileNameAndSwagger struct {
+	fileName string
+	swagger  spec.Swagger
+}
+
+// loadFileWithoutCache loads the schema at the specified file path. it does not read from
+// or add to the cache.
+func (fileCache *OpenAPISchemaCache) loadUncachedFile(filePath string) (spec.Swagger, error) {
+	var swagger spec.Swagger
+
+	fileContent, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return swagger, errors.Wrap(err, "unable to read swagger file")
+	}
+
+	err = swagger.UnmarshalJSON(fileContent)
+	if err != nil {
+		return swagger, errors.Wrap(err, "unable to parse swagger file")
+	}
+
+	return swagger, err
+}
+
+// PreloadCache loads the specified schema into the cache. it does not attempt to deduplicate work.
+func (fileCache *OpenAPISchemaCache) PreloadCache(filePath string) (spec.Swagger, error) {
```

Comment:
what's the advantage of having this and `fetchFileAbsolute`? I know this is exported and `fetchFileAbsolute` isn't, and `fetchFileAbsolute` has double-checked locking/is threadsafe, but that just makes `fetchFileAbsolute` a superset of this?

Would it make more sense to just expose `fetchFileAbsolute` and remove `PreloadCache`?

Reply:
`fetchFileAbsolute` only lets one file be loaded at any one time, this will load as many in parallel as you want (but would duplicate work if you load duplicate filenames).

The threadsafety here is not actually needed any more since I pulled the type-extraction out as its own pass over the schema types, so I can simplify this type greatly.

Reply:
Stripped out most of this.